autorag/nodes/retrieval/bm25.py

```diff
@@ -94,18 +95,48 @@ def bm25_ingest(corpus_path: str, corpus_data: pd.DataFrame):
     if not corpus_path.endswith('.pkl'):
         raise ValueError(f"Corpus path {corpus_path} is not a pickle file.")
     validate_corpus_dataset(corpus_data)
-    tokenizer = AutoTokenizer.from_pretrained('gpt2', use_fast=False)
-    tasks = list(
-        map(lambda x: bm25_tokenize(x[0], x[1], tokenizer), zip(corpus_data['contents'], corpus_data['doc_id'])))
-    loop = asyncio.get_event_loop()
-    results = loop.run_until_complete(asyncio.gather(*tasks))
-    tokenized_corpus, passage_ids = zip(*results)
-    bm25_dict = {
-        'tokens': list(tokenized_corpus),
-        'passage_id': list(passage_ids),
-    }
-    with open(corpus_path, 'wb') as w:
-        pickle.dump(bm25_dict, w)
+
+    ids = corpus_data['doc_id'].tolist()
+    contents = corpus_data['contents'].tolist()
+
+    # Initialize bm25_corpus
+    bm25_corpus = pd.DataFrame()
+
+    # Load the BM25 corpus if it exists and get the passage ids
+    if os.path.exists(corpus_path) and os.path.getsize(corpus_path) > 0:
+        with open(corpus_path, 'rb') as r:
+            corpus = pickle.load(r)
+            bm25_corpus = pd.DataFrame.from_dict(corpus)
+        stored_passage_rows = bm25_corpus[bm25_corpus['passage_id'].isin(ids)]
+        stored_passage_ids = stored_passage_rows['passage_id'].tolist()
+        # Filter ids and contents for those not existing in the BM25 corpus
+        new_ids = [passage_id for passage_id in ids if passage_id not in stored_passage_ids]
+        new_contents = [contents[i] for i, passage_id in enumerate(ids) if passage_id in new_ids]
+    else:
+        new_ids = ids
+        new_contents = contents
+
+    if new_ids:
+        tokenizer = AutoTokenizer.from_pretrained('gpt2', use_fast=False)
+        tasks = list(
+            map(lambda x: bm25_tokenize(x[0], x[1], tokenizer), zip(new_contents, new_ids)))
+        loop = asyncio.get_event_loop()
+        results = loop.run_until_complete(asyncio.gather(*tasks))
+        tokenized_corpus, passage_ids = zip(*results)
+
+        new_bm25_corpus = pd.DataFrame({
```

Comment:
If you re-make `new_bm25_corpus` again, why did you use `new_ids` and `new_contents` separately?

First, make bm25.pkl dict to pd.DataFrame.
Second, do tokenize new ids. 
You can apply with async using `swifter`.
Third, concat it!